bolt_mujoco_simulation/src/mujoco_simulator.cpp

```diff
@@ -285,11 +335,63 @@ void MuJoCoSimulator::write(const std::map<std::string, double> &pos,
 void MuJoCoSimulator::syncStates() {
   for (auto i = 0; i < m->nu; ++i) {
     std::string name = m->names + m->name_actuatoradr[i];
-    
+
     pos_state[name] = d->qpos[i + freeflyer_nq];
     vel_state[name] = d->actuator_velocity[i];
     eff_state[name] = d->actuator_force[i];
   }
+
+  // sensor_data
+  sensor_orientation = {
+    d->sensordata[sensor_orient_offset],
+    d->sensordata[sensor_orient_offset + 1],
+    d->sensordata[sensor_orient_offset + 2],
+    d->sensordata[sensor_orient_offset + 3],
+  };
+  sensor_angular_vel = {
+    d->sensordata[sensor_angular_vel_offset],
+    d->sensordata[sensor_angular_vel_offset + 1],
+    d->sensordata[sensor_angular_vel_offset + 2]
+  };
+  sensor_linear_acc = {
+    d->sensordata[sensor_linear_acc_offset],
+    d->sensordata[sensor_linear_acc_offset + 1],
+    d->sensordata[sensor_linear_acc_offset + 2]
+  };
+}
+
+void MuJoCoSimulator::setNoiseParams(double mean, double dev) {
+  std::random_device rd;
+  random_engine = std::mt19937(rd());
+  normal_dist = std::normal_distribution<double>(mean, dev);
+}
+
+double MuJoCoSimulator::computeNoise() {
+  if (!enable_noise) {
+    return 0.0;
+  }
+
+  return normal_dist(random_engine);
 }
 
+void MuJoCoSimulator::publishImuData() {
+  auto now = node->get_clock()->now();
+  sensor_msgs::msg::Imu msg;
+  msg.header.frame_id = NAME_SENSOR_IMU_LINK;
+  msg.header.stamp.sec = now.seconds();
+  msg.header.stamp.nanosec = now.nanoseconds();
+
+  msg.angular_velocity.x = computeNoise() + d->sensordata[sensor_angular_vel_offset];
+  msg.angular_velocity.y = computeNoise() + d->sensordata[sensor_angular_vel_offset + 1];
+  msg.angular_velocity.z = computeNoise() + d->sensordata[sensor_angular_vel_offset + 2];
+  msg.linear_acceleration.x = computeNoise() + d->sensordata[sensor_linear_acc_offset];
+  msg.linear_acceleration.y = computeNoise() + d->sensordata[sensor_linear_acc_offset + 1];
+  msg.linear_acceleration.z = computeNoise() + d->sensordata[sensor_linear_acc_offset + 2];
+  msg.orientation.w = computeNoise() + d->sensordata[sensor_orient_offset];
+  msg.orientation.x = computeNoise() + d->sensordata[sensor_orient_offset + 1];
+  msg.orientation.y = computeNoise() + d->sensordata[sensor_orient_offset + 2];
+  msg.orientation.z = computeNoise() + d->sensordata[sensor_orient_offset + 3];
```

Comment:
I very much dislike this. This looks like the noise is applied all the time, one has to check the `computeNoise` function to see how its actually handled. The check whether or not noise is added should occur here.